File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();
        
        for (int i = 0; i < nums.length; i++){
            if (i == 0 || nums[i-1] != nums[i]){
                threeSum(nums, i, ans);
            }
        }
        return ans;
        
    }
    
    void threeSum(int[] nums, int i, List<List<Integer>> ans){
        int l = i + 1, h = nums.length -1;
        while (l < h){
            int sum = nums[i] + nums[l] + nums[h];
            if (sum < 0){
                ++l;
            }
            else if (sum > 0){
                --h;
            }
            else{
                ans.add(Arrays.asList(nums[i], nums[l], nums[h]));
                ++l;
                //--h;
                while (l < h && nums[l-1] == nums[l]){
                    ++l;
                }
            }
        }
    }
}

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        
        for i, curr in enumerate(nums):
            if i > 0 and curr == nums[i-1]:
                continue
                
            l, r = i + 1, len(nums) - 1
            
            while (l < r):
                three_sum = curr + nums[l] + nums[r]
                if three_sum > 0:
                    r -= 1
                elif three_sum < 0:
                    l += 1
                else:
                    res.append([curr, nums[l], nums[r]])
                    l += 1
                    while l < r and nums[l-1] == nums[l]:
                        l += 1
        
        return res
                